core/sessions/orm.go

```diff
@@ -41,66 +45,106 @@ type ORM interface {
+// FindUserByAPIToken will attempt to return an API user via the user's table token_key column.
+func (o *orm) FindUserByAPIToken(apiToken string) (user User, err error) {
+	sql := "SELECT * FROM users WHERE token_key = $1"
+	err = o.q.Get(&user, sql, apiToken)
 	return
 }
 
-// AuthorizedUserWithSession will return the one API user if the Session ID exists
-// and hasn't expired, and update session's LastUsed field.
+func (o *orm) findUser(email string) (user User, err error) {
+	sql := "SELECT * FROM users WHERE lower(email) = lower($1)"
+	err = o.q.Get(&user, sql, email)
+	return
+}
+
+// ListUsers will load and return all user rows from the db.
+func (o *orm) ListUsers() (users []User, err error) {
+	sql := "SELECT * FROM users ORDER BY email ASC;"
+	err = o.q.Select(&users, sql)
+	return
+}
+
+// AuthorizedUserWithSession will return the API user associated with the Session ID if it
+// exists and hasn't expired, and update session's LastUsed field.
 func (o *orm) AuthorizedUserWithSession(sessionID string) (User, error) {
 	if len(sessionID) == 0 {
 		return User{}, errors.New("Session ID cannot be empty")
 	}
 
-	result, err := o.db.Exec("UPDATE sessions SET last_used = now() WHERE id = $1 AND last_used + $2 >= now()", sessionID, o.sessionDuration)
-	if err != nil {
-		return User{}, err
-	}
-	rowsAffected, err := result.RowsAffected()
+	var user User
+	err := o.q.Transaction(func(tx pg.Queryer) error {
+		// First find user based on session token
+		var email string
+		if err := tx.Get(&email, "SELECT email FROM sessions WHERE id = $1 FOR UPDATE", sessionID); err != nil {
+			return errors.Wrap(err, "no matching user for provided session token")
+		}
+		if err := tx.Get(&user, "SELECT * FROM users WHERE lower(email) = lower($1)", email); err != nil {
+			return errors.Wrap(err, "no matching user for provided session email")
+		}
+		// Sesssion valid and tied to user, update last_used
+		result, err := tx.Exec("UPDATE sessions SET last_used = now() WHERE id = $1 AND last_used + $2 >= now()", sessionID, o.sessionDuration)
+		if err != nil {
+			return errors.Wrap(err, "unable to update sessions table")
+		}
+		rowsAffected, err := result.RowsAffected()
+		if err != nil {
+			return err
+		}
+		if rowsAffected == 0 {
+			return sql.ErrNoRows
+		}
+		return nil
+	})
+
 	if err != nil {
 		return User{}, err
 	}
-	if rowsAffected == 0 {
-		return User{}, sql.ErrNoRows
-	}
-	return o.FindUser()
+
+	return user, nil
 }
 
-// DeleteUser will delete the API User in the db.
-func (o *orm) DeleteUser() error {
-	ctx, cancel := pg.DefaultQueryCtx()
-	defer cancel()
-	return pg.SqlxTransaction(ctx, o.db, o.lggr, func(tx pg.Queryer) error {
-		if _, err := tx.Exec("DELETE FROM users"); err != nil {
+// DeleteUser will delete an API User and sessions by email.
+func (o *orm) DeleteUser(email string) error {
+	return o.q.Transaction(func(tx pg.Queryer) error {
+		// session table rows are deleted on cascade through the user email constraint
```

Comment:
Is there a test case covering this?

Reply:
That's a good point, I've added a test case now 🙂